crates/analytics/src/lib.rs

```diff
@@ -0,0 +1,229 @@
+use async_trait::async_trait;
+use std::ops::Div;
+
+use rain_orderbook_subgraph_client::types::common::Trade;
+use rain_orderbook_subgraph_client::{OrderbookSubgraphClient, OrderbookSubgraphClientError};
+
+#[async_trait]
+pub trait OrderbookSubgraphClientTrait {
+    async fn all_trades_list(
+        &self,
+        timestamp_gte: Option<u64>,
+        timestamp_lte: Option<u64>,
+    ) -> Result<Vec<Trade>, OrderbookSubgraphClientError>;
+}
+
+#[async_trait]
+impl OrderbookSubgraphClientTrait for OrderbookSubgraphClient {
+    async fn all_trades_list(
+        &self,
+        timestamp_gte: Option<u64>,
+        timestamp_lte: Option<u64>,
+    ) -> Result<Vec<Trade>, OrderbookSubgraphClientError> {
+        self.all_trades_list(timestamp_gte, timestamp_lte).await
+    }
+}
+
+pub struct Analytics<T: OrderbookSubgraphClientTrait + Send + Sync> {
+    client: T,
+}
+
+impl<T: OrderbookSubgraphClientTrait + Send + Sync> Analytics<T> {
+    pub fn new(client: T) -> Self {
+        Self { client }
+    }
+
+    /// Downtime Metrics based on time between trades
+    pub async fn calculate_downtime_between_trades(
+        &self,
+        period: Option<(u64, u64)>,
+    ) -> (f64, f64, f64) {
+        let trades: Vec<Trade> = match period {
+            Some((start, end)) => self.client.all_trades_list(Some(start), Some(end)).await,
+            None => self.client.all_trades_list(None, None).await,
+        }
+        .unwrap_or_default();
+
+        let mut time_diffs: Vec<u64> = Vec::new();
+        for window in trades.windows(2) {
+            if let [prev, curr] = window {
+                let diff = curr.timestamp.0.parse::<u64>().unwrap()
+                    - prev.timestamp.0.parse::<u64>().unwrap();
+                time_diffs.push(diff);
+            }
+        }
+
+        if time_diffs.is_empty() {
+            return (0.0, 0.0, 0.0);
+        }
+
+        let avg: f64 = time_diffs.iter().sum::<u64>().div(time_diffs.len() as u64) as f64;
+        let min: f64 = *time_diffs.iter().min().unwrap() as f64;
+        let max: f64 = *time_diffs.iter().max().unwrap() as f64;
+
+        (avg, min, max)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rain_orderbook_subgraph_client::types::common::{
+        BigInt, Bytes, Erc20, Orderbook, Trade, TradeEvent, TradeStructPartialOrder,
+        TradeVaultBalanceChange, Transaction, VaultBalanceChangeVault,
+    };
+
+    struct MockSubgraphClient {
+        trades: Vec<Trade>,
+    }
+
+    #[async_trait::async_trait]
+    impl OrderbookSubgraphClientTrait for MockSubgraphClient {
+        async fn all_trades_list(
+            &self,
+            _timestamp_gte: Option<u64>,
+            _timestamp_lte: Option<u64>,
+        ) -> Result<Vec<Trade>, OrderbookSubgraphClientError> {
+            Ok(self.trades.clone())
+        }
+    }
+
+    fn create_mock_trade(timestamp: u64) -> Trade {
+        Trade {
```

Comment:
Note: IMO this should had been provided as a mock interface from the subgraph crate.

Reply:
makes sense - can you move it over there?

Reply:
I can, but this particular block is very ugly and opinionated for this set of unit tests. Also out of scope of this PR i'd say. 

What I's thinking was some sort of a testing crate that creates fake trades & orders for a given scenario.